agentstack/generation/tool_generation.py

```diff
@@ -9,65 +9,130 @@
 import shutil
 import fileinput
 
+TOOL_INIT_FILENAME = "src/tools/__init__.py"
+AGENTSTACK_JSON_FILENAME = "agentstack.json"
+
 
 def add_tool(tool_name: str, path: Optional[str] = None):
+    if path:
+        path = path.endswith('/') and path or path + '/'
+    else:
+        path = './'
     with importlib.resources.path(f'agentstack.tools', 'tools.json') as tools_data_path:
         tools = open_json_file(tools_data_path)
         framework = get_framework(path)
         assert_tool_exists(tool_name, tools)
+        agentstack_json = open_json_file(f'{path}{AGENTSTACK_JSON_FILENAME}')
+        
+        if tool_name in agentstack_json.get('tools', []):
+            print(term_color(f'Tool {tool_name} is already installed', 'red'))
+            sys.exit(1)
 
         with importlib.resources.path(f'agentstack.tools', f"{tool_name}.json") as tool_data_path:
             tool_data = open_json_file(tool_data_path)
 
             with importlib.resources.path(f'agentstack.templates.{framework}.tools', f"{tool_name}_tool.py") as tool_file_path:
-                os.system(tool_data['package'])  # Install package
-                shutil.copy(tool_file_path, f'{path + "/" if path else ""}src/tools/{tool_name}_tool.py')  # Move tool from package to project
+                if tool_data.get('packages'):
+                    os.system(f'poetry add {' '.join(tool_data["packages"])}')  # Install packages
```

Comment:
this fails in python 3.11 because string primitives cant be used in f-strings (annoying ik)

```python
              if tool_data.get('packages'):
                  packages = ' '.join(tool_data["packages"])
                  os.system(f'poetry add {packages}')  # Install packages
```

Reply:
actually, python didn't complain about it on line 74, so maybe its just because it's a single quote nested inside a single quote. try this first:

```python
if tool_data.get('packages'):
    os.system(f'poetry add {" ".join(tool_data["packages"])}')
```

Reply:
oh, yeah, that's totally what it was. latest commit puts them back on one line

Reply:
hell yeah. thanks!!!